File: packages/api/src/infrastructure/transports/presets.ts
import { createCommandTransport } from "./command";
import type { AITransport } from "../../ports/provider";

export const agentPresetIds = ["claude-code", "opencode", "codex", "crush", "gemini"] as const;
export type AgentPreset = (typeof agentPresetIds)[number];

interface PresetDefinition {
  command: string;
  args: string[];
  promptVia?: "stdin" | "argv";
}

/**
 * Command + args per preset, each carrying the flag that stops the agent
 * from blocking an unattended build on a permission prompt. Re-verified
 * against installed CLIs (not just published docs) on 2026-08-11:
 * - opencode: `--yolo` does not exist on this CLI; the real flag is `--auto`.
 * - codex: `--approve-for-me` cannot be combined with `--sandbox`.
 * - crush: `crush run [prompt...]` takes the prompt as argv tokens, not
 *   stdin (stdin is documented only for supplementary piped content), so
 *   this preset sets `promptVia: "argv"`.
 * Agent CLI surfaces drift fast — re-verify before trusting this table.
 */
const presets: Record<AgentPreset, PresetDefinition> = {
  "claude-code": {
    command: "claude",
    args: ["-p", "--dangerously-skip-permissions", "--output-format", "text"],
  },
  opencode: { command: "opencode", args: ["run", "--auto"] },
  codex: { command: "codex", args: ["exec", "--approve-for-me"] },
  crush: { command: "crush", args: ["run", "-q"], promptVia: "argv" },
  // gemini's -p takes a value and appends stdin to it; pass "" so the full
  // prompt travels on stdin like every other preset.
  gemini: { command: "gemini", args: ["-p", "", "-y"] },
};

export function resolveAgentPreset(preset: AgentPreset, cwd?: string): AITransport {
  const def = presets[preset];
  return createCommandTransport({
    id: preset,
    command: def.command,
    args: def.args,
    cwd,
    promptVia: def.promptVia,
  });
}

File: crates/intl-ai-providers/src/presets.rs
//! Agent preset table, ported verbatim from `transports/presets.ts`
//! (plan 5.1.7 — the args lists are the contract).

use intl_ai_core::config::{AgentPreset, PromptVia};

pub struct Preset {
    pub command: &'static str,
    pub args: &'static [&'static str],
    pub prompt_via: PromptVia,
}

pub fn resolve(preset: AgentPreset) -> Preset {
    match preset {
        AgentPreset::ClaudeCode => Preset {
            command: "claude",
            args: &[
                "-p",
                "--dangerously-skip-permissions",
                "--output-format",
                "text",
            ],
            prompt_via: PromptVia::Stdin,
        },
        AgentPreset::Opencode => Preset {
            command: "opencode",
            args: &["run", "--auto"],
            prompt_via: PromptVia::Stdin,
        },
        AgentPreset::Codex => Preset {
            command: "codex",
            args: &["exec", "--approve-for-me"],
            prompt_via: PromptVia::Stdin,
        },
        AgentPreset::Crush => Preset {
            command: "crush",
            args: &["run", "-q"],
            prompt_via: PromptVia::Argv,
        },
        AgentPreset::Gemini => Preset {
            command: "gemini",
            args: &["-p", "", "-y"],
            prompt_via: PromptVia::Stdin,
        },
    }
}

pub fn preset_id(preset: AgentPreset) -> &'static str {
    match preset {
        AgentPreset::ClaudeCode => "claude-code",
        AgentPreset::Opencode => "opencode",
        AgentPreset::Codex => "codex",
        AgentPreset::Crush => "crush",
        AgentPreset::Gemini => "gemini",
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn presets_match_ts_table() {
        assert_eq!(
            resolve(AgentPreset::ClaudeCode).args,
            &[
                "-p",
                "--dangerously-skip-permissions",
                "--output-format",
                "text"
            ]
        );
        assert_eq!(resolve(AgentPreset::Opencode).args, &["run", "--auto"]);
        assert_eq!(
            resolve(AgentPreset::Codex).args,
            &["exec", "--approve-for-me"]
        );
        assert_eq!(resolve(AgentPreset::Crush).args, &["run", "-q"]);
        assert!(matches!(
            resolve(AgentPreset::Crush).prompt_via,
            PromptVia::Argv
        ));
        assert_eq!(resolve(AgentPreset::Gemini).args, &["-p", "", "-y"]);
        assert!(matches!(
            resolve(AgentPreset::Gemini).prompt_via,
            PromptVia::Stdin
        ));
    }
}
